src/adera/ShipResources.cpp

```diff
@@ -0,0 +1,119 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "ShipResources.h"
+#include "osp/Active/ActiveScene.h"
+
+using namespace osp;
+using namespace osp::active;
+using namespace adera::active::machines;
+
+/* MachineContainer */
+
+void MachineContainer::propagate_output(WireOutput* output)
+{
+
+}
+
+WireInput* MachineContainer::request_input(WireInPort port)
+{
+    return nullptr;
+}
+
+WireOutput* MachineContainer::request_output(WireOutPort port)
+{
+    return nullptr;
+}
+
+std::vector<WireInput*> MachineContainer::existing_inputs()
+{
+    return m_inputs;
+}
+
+std::vector<WireOutput*> MachineContainer::existing_outputs()
+{
+    return m_outputs;
+}
+
+/* SysMachineContainer */
+
+SysMachineContainer::SysMachineContainer(ActiveScene& rScene)
+    : SysMachine<SysMachineContainer, MachineContainer>(rScene)
+    , m_updateContainers(rScene.get_update_order(), "mach_container", "", "mach_rocket",
+        [this](ActiveScene& rScene) { this->update_containers(rScene); })
+{ }
+
+void SysMachineContainer::update_containers(ActiveScene& rScene)
+{
+    auto view = rScene.get_registry().view<MachineContainer>();
+
+    for (ActiveEnt ent : view)
+    {
+        auto& container = view.get<MachineContainer>(ent);
+
+        // Do something useful here... or not
+    }
+}
+
+float SysMachineContainer::resource_volume(ShipResourceType type, uint64_t quantity)
+{
+    uint64_t units = quantity / (1ull << type.m_quanta);
```

Comment:
Your intention here is to multiply 1 by 2, m_quanta times?

If I recall correctly, std::pow is floating point only (stupid).

I recommend making a utility function that abstracts the bit-shifting. Makes it much easier for future readers to easily understand what you're trying to accomplish.

Or, considering you're always casting to float anyway, maybe just use std::pow and do everything as float.

Reply:
std::pow will work. I'm currently wondering how these functions worked at all since 1/(large int) is zero, so I'll need to take a closer look with my next PR. Maybe I fix them in the later implementation.